docs/execution.md

```diff
@@ -0,0 +1,208 @@
+* a slot is SCE-final if its immediate predecessor slot contains a block that is CSS-final or is empty but followed in its own thread by a CSS-final block. This ensures that no new block executions can be inserted in-between existing final block executions in the SCE
+
+When a block B becomes SCE-final:
+* update the Final SCE ledger by VM-executing the newly SCE-final block on top of the current Final SCE ledger 
+* if the block B requires transferring coins from the SCE ledger to the CSS ledeger, send a TransferToConsensus to the CSS:
+  * fields:
+    * target_addr = the emitter of the operation that caused the TransferToConsensus
+    * origin_slot = the slot of the SCE-final block causing the TransferToConsensus
+    * amount = amount of coins to transfer
+  * this will cause the spending of "amount" coins on the SCE side
+
+When the CSS receives a TransferToConsensus message:
+* it registers that `target_addr` needs to be credited of `amount` coins in the CSS ledger at slot `target_slot = origin_slot + 1 period`
+* the credit is performed at the beginning of the first block that comes at `target_slot` (or later in case of misses)
+
+
+### Structure of the SCE
+
+
+```mermaid
+flowchart TB
+    Address-->Bytecode
+    subgraph SCE Ledger Data
+      Address-->Balance
+      Address-->DB
+    end
+    subgraph Program
+      Bytecode
+    end
+```
+
+* Holds separate Final and Candidate SCE ledgers, that match each address to a balance, database, and program area
+  * the `database` is a key-value map where the key is a sha256 hash, and the value contains arbitrary bytes
+  * the `program` area contains bytecode with decorated public/private functions
+* Contains an execution engine that will be fed the sorted and ExecuteSC operations with the block they belong to:
+  * Bytecode is interpreted as WASM and executed
+  * massa-specific "systcalls" are made available to the execution engine
+  * the execution engine is fully deterministic
+  * the execution engine measures and limits resource and gas usage during realtime during bytecode execution
+
+* the SCE ledger is basically a `HashMap<Address, SCELedgerEntry>` where SCELedgerEntry has the fields:
+  * sce_balance: `Amount`
+  * database: `HashMap<Hash, Vec<u8>>`
+  * program_data: `Vec<u8>` // to be better defined in followup issues
+
+### SCE WASM execution engine syscalls
+
+A smart-contract will be able to call into Massa-specific API's, providing by the "Massa runtime".
+We could use "host functions" in wasmer to provide these API's to a running contract. See https://gitlab.com/massalabs/massa/-/issues/360
+
+Available "syscalls":
+
+- `Call(SC_addr, function_name, parameters)`: synchronously calls a function in the same, or another smart contract
+- `TransferToConsensus(amount)`: transfers massa coins from SCE ledger to the consensus ledger. On the SCE side, this just burns the coins.
+- `WriteDatabase(key, value)`: writes in the database of the current smart contract
```

Comment:
Rights are checked inside that write function ?

Reply:
this is not a final list, just a preliminary idea.

But if we go the way `WriteDatabase(key, value)`, there is no rights check to be done because it is always targeted at the current contect, which always has write rights to itself